File: blink-macos/BlinkCore/Sources/BlinkCore/Timer/AdaptiveTimingEngine.swift
import Foundation

/// Learns the user's natural break rhythm and adjusts timer duration.
/// Analyzes which break intervals the user actually complies with.
public final class AdaptiveTimingEngine {
    private var acceptedIntervals: [TimeInterval] = []
    private static let maxHistory = 50
    private static let minInterval: TimeInterval = 900   // 15 minutes
    private static let maxInterval: TimeInterval = 2700  // 45 minutes

    public init() {}

    /// Record that a break was accepted at this interval since last break.
    public func recordAcceptedBreak(intervalSinceLastBreak: TimeInterval) {
        acceptedIntervals.append(intervalSinceLastBreak)
        if acceptedIntervals.count > Self.maxHistory {
            acceptedIntervals.removeFirst()
        }
    }

    /// Returns the suggested base timer duration based on learned behavior.
    /// Returns nil if not enough data to make a recommendation.
    public func suggestedDuration() -> TimeInterval? {
        guard acceptedIntervals.count >= 10 else { return nil }

        // Use the median of recent accepted intervals
        let sorted = acceptedIntervals.sorted()
        let median = sorted[sorted.count / 2]

        return min(Self.maxInterval, max(Self.minInterval, median))
    }

    /// Load saved intervals (from persistence).
    public func load(intervals: [TimeInterval]) {
        acceptedIntervals = Array(intervals.suffix(Self.maxHistory))
    }

    /// Current intervals for persistence.
    public func savedIntervals() -> [TimeInterval] {
        acceptedIntervals
    }
}

File: blink-windows/src/Blink.Core/Timer/AdaptiveTimingEngine.cs
namespace Blink.Core.Timer;

public sealed class AdaptiveTimingEngine
{
    private readonly List<double> _acceptedIntervals = [];
    private const int MaxHistory = 50;
    private const double MinInterval = 900;   // 15 minutes
    private const double MaxInterval = 2700;  // 45 minutes

    public void RecordAcceptedBreak(double intervalSinceLastBreak)
    {
        _acceptedIntervals.Add(intervalSinceLastBreak);
        if (_acceptedIntervals.Count > MaxHistory)
            _acceptedIntervals.RemoveAt(0);
    }

    public double? SuggestedDuration()
    {
        if (_acceptedIntervals.Count < 10) return null;

        var sorted = _acceptedIntervals.OrderBy(x => x).ToList();
        var median = sorted[sorted.Count / 2];

        return Math.Min(MaxInterval, Math.Max(MinInterval, median));
    }

    public void Load(double[] intervals)
    {
        _acceptedIntervals.Clear();
        _acceptedIntervals.AddRange(intervals.TakeLast(MaxHistory));
    }

    public double[] SavedIntervals() => [.. _acceptedIntervals];
}
